benchmark/_driver.py

```diff
@@ -0,0 +1,139 @@
+        If ``success is False``, the dictionary also contains a ``reason`` for
+        failure.
+    """
+    scenario_established = scenario.start()
+
+    samples = []
+
+    def sample(i):
+        with start_action(action_type=u'flocker:benchmark:sample', sample=i):
+            sampling = DeferredContext(maybeDeferred(operation.get_probe))
+
+            def run_probe(probe):
+                probing = metric.measure(probe.run)
+                probing.addCallbacks(
+                    lambda interval: samples.append(
+                        dict(success=True, value=interval)
+                    ),
+                    lambda reason: samples.append(
+                        dict(success=False, reason=reason.getTraceback()),
+                    ),
+                )
+                probing.addCallback(lambda ignored: probe.cleanup())
+                return probing
+            sampling.addCallback(run_probe)
+            sampling.addActionFinish()
+            return sampling.result
+
+    def collect_samples(ignored):
+        collecting = Deferred()
+        task = cooperate(sample(i) for i in range(num_samples))
+
+        # If the scenario collapses, stop sampling
+        def stop_sampling_on_scenario_collapse(failure):
+            task.stop()
+            collecting.errback(failure)
+        scenario.maintained().addErrback(stop_sampling_on_scenario_collapse)
+
+        # Leaving the errback unhandled makes tests fail
+        task.whenDone().addCallbacks(
+            lambda ignored: collecting.callback(samples),
+            lambda ignored: None)
+
+        return collecting
+
+    benchmarking = scenario_established.addCallback(collect_samples)
+
+    def tear_down(result):
+        stopping = scenario.stop()
+        stopping.addCallback(lambda ignored: result)
+        return stopping
+
+    benchmarking.addBoth(tear_down)
+
+    return benchmarking
+
+
+def driver(reactor, config, operation, metric, scenario, result, output):
+    """
+    :param reactor: Reactor to use.
+    :param config: Configuration read from options.
+    :param IOperation operation: An operation to perform.
+    :param IMetric metric: A quantity to measure.
+    :param IScenario scenario: A load scenario.
+    :param result: A dictionary which will be updated with values to
+        create a JSON result.
+    :param output: A callable to receive the JSON structure, for
+        printing or storage.
+    """
+
+    if config['control']:
+        cert_directory = FilePath(config['certs'])
+        control_service = FlockerClient(
+            reactor,
+            host=config['control'],
+            port=4523,
+            ca_cluster_path=cert_directory.child(b"cluster.crt"),
+            cert_path=cert_directory.child(b"user.crt"),
+            key_path=cert_directory.child(b"user.key"),
+        )
+
+        d = control_service.version()
+    else:
+        # Only valid for operation 'no-op'
```

Comment:
Yes. I found that it was very easy to leave off the `control` option, and not realise that you are not testing a real cluster, but it is useful to be able to run the benchmark without a cluster.  Since benchmarking `no-op` makes no real sense, you must be testing in this case.

A better solution might be to have an explicit name for a fake cluster (e.g. `--control=test`)?